platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/AbstractResourcePattern.java

```diff
@@ -59,7 +59,9 @@ private AbstractResourcePattern(ExternalResourceName pattern) {
 
     @Override
     public String getPattern() {
-        return pattern.getDecoded();
+        // Replace encoded [] with plain variants, since we are using them with a special meaning here.
+        // Inefficient but easy to read, I don't think this method is hot.
+        return pattern.getDisplayable().replace("%5B", "[").replace("%5D", "]");
```

Comment:
The encoded `[]` come from the normal URI encoding rules, which state that certain characters are reserved for having special meaning in schemes and must be encoded; see `gen-delims` in [RFC 3986 Section 2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2). Since in the pattern, these actually do have a special meaning, I decode them here so that patterns are not displayed with encoded `[]`.

Reply:
I benchmarked this against the loop I would've written, and it turns out that the `replace` calls are actually faster!
```
Benchmark                                                                                                               (rawUri)  Mode  Cnt    Score    Error  Units
UriPerf.decodeUriReplace                            https://foo.bar/repo/[organization]/[revision]/[module](.[classifier]).[ext]  avgt   10  127.136 ±  8.467  ns/op
UriPerf.decodeUriReplace      https://foo.bar/repo/[organization]/[revision]/[module]/[artifact]-[revision](.[classifier]).[ext]  avgt   10  165.374 ± 16.068  ns/op
UriPerf.decodeUriReplaceLoop                        https://foo.bar/repo/[organization]/[revision]/[module](.[classifier]).[ext]  avgt   10  216.316 ±  9.639  ns/op
UriPerf.decodeUriReplaceLoop  https://foo.bar/repo/[organization]/[revision]/[module]/[artifact]-[revision](.[classifier]).[ext]  avgt   10  281.815 ±  7.584  ns/op
```
So I'll remove the efficiency comment, as I assumed incorrectly.